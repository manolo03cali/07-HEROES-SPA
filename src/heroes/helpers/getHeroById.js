import { heroes } from "../data/heroes";
export const getHeroyById = (id) => {
  return heroes.find((hero) => hero.id === id);
};
