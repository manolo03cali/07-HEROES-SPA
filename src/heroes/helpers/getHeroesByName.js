// Importa el arreglo de héroes desde el archivo "../data/heroes"
import { heroes } from "../data/heroes";

// Exporta una función llamada getHeroesByName que recibe un parámetro 'name'
// Si no se pasa ningún valor, 'name' será una cadena vacía por defecto
export const getHeroesByName = (name = "") => {
  // Convierte 'name' a minúsculas y elimina espacios en blanco al inicio y al final
  name = name.toLocaleLowerCase().trim();

  // Si la cadena 'name' está vacía después de limpiar, retorna un arreglo vacío
  if (name.length === 0) return [];

  // Filtra el arreglo 'heroes' para encontrar los héroes cuyo nombre ('superhero')
  // incluya la cadena 'name'. También convierte cada nombre a minúsculas
  // para que la búsqueda no sea sensible a mayúsculas/minúsculas
  return heroes.filter((hero) =>
    hero.superhero.toLocaleLowerCase().includes(name)
  );
};
