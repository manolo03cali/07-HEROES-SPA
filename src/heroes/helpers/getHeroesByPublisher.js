// Importamos el arreglo 'heroes' desde el archivo '../data/heroes'
import { heroes } from "../data/heroes";

// Exportamos una función llamada 'getHeroesByPublisher' que recibe como parámetro 'publisher'
export const getHeroesByPublisher = (publisher) => {
  // Definimos un arreglo con los nombres de los publishers válidos
  const validPublishers = ["DC Comics", "Marvel Comics"];

  // Verificamos si el publisher recibido no está en la lista de publishers válidos
  if (!validPublishers.includes(publisher)) {
    // Si no es válido, lanzamos un error indicando que el publisher no es válido
    throw new Error(`${publisher} is no valid publisher`);
  }

  // Si el publisher es válido, filtramos el arreglo de 'heroes'
  // Devolvemos un nuevo arreglo solo con los héroes cuyo 'publisher' sea igual al recibido
  return heroes.filter((heroe) => heroe.publisher === publisher);
};
