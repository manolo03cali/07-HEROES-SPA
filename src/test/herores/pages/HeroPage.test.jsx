// Importación de herramientas de testing
import { render, screen, fireEvent } from "@testing-library/react"; // Funciones básicas para renderizar y probar componentes
import { MemoryRouter, Routes, Route, useParams } from "react-router-dom"; // Componentes para simular rutas en pruebas
import { HeroPage } from "../../../heroes"; // Componente que vamos a probar
import { getHeroyById } from "../../../heroes/helpers"; // Función para obtener datos de héroes
import "@testing-library/jest-dom"; // Extiende Jest con matchers para el DOM

// Mock de useNavigate para simular navegación
const mockUseNavigate = jest.fn(); // Creamos una función mock para simular useNavigate

// Configuración de mocks para react-router-dom
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"), // Mantenemos todas las funcionalidades reales
  useNavigate: () => mockUseNavigate, // Reemplazamos useNavigate con nuestro mock
  useParams: jest.fn(), // Mock de useParams para controlar los parámetros de ruta
}));

// Bloque principal de pruebas para HeroPage
describe("Pruebas en <HeroPage />", () => {
  // Configuración que se ejecuta antes de cada prueba
  beforeEach(() => {
    jest.clearAllMocks(); // Limpiamos todos los mocks para evitar interferencias
    // Configuramos useParams para devolver un ID por defecto (Batman)
    useParams.mockReturnValue({ id: "dc-batman" });
  });

  // Prueba de Snapshot: Verifica que el componente se renderice correctamente
  test("Debe de mostrarse correctamente Snapshot", () => {
    // Renderizamos el componente dentro de MemoryRouter
    const { container } = render(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Routes>
          <Route path="/hero/:id" element={<HeroPage />} />
        </Routes>
      </MemoryRouter>
    );
    // Comparamos con el snapshot guardado
    expect(container).toMatchSnapshot();
  });

  // Prueba 1: Verifica el renderizado correcto cuando el héroe existe
  test("debe mostrar el héroe correctamente cuando existe", () => {
    // Obtenemos los datos reales del héroe usando el helper
    const realHero = getHeroyById("dc-batman");

    // Renderizamos el componente con la ruta inicial
    render(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Routes>
          <Route path="/hero/:id" element={<HeroPage />} />
        </Routes>
      </MemoryRouter>
    );

    /* Verificaciones del contenido renderizado */

    // 1. Verificamos que el nombre del superhéroe se muestra
    expect(screen.getByText(realHero.superhero)).toBeInTheDocument();

    // 2. Verificación del alter ego (buscamos el texto y subimos al elemento li padre)
    const alterEgoItem = screen.getByText("Alter ego:").closest("li");
    expect(alterEgoItem).toHaveTextContent(realHero.alter_ego);

    // 3. Verificamos otros datos del héroe
    expect(screen.getByText(realHero.publisher)).toBeInTheDocument();
    expect(screen.getByText(realHero.first_appearance)).toBeInTheDocument();

    // 4. Verificación de la imagen del héroe
    const img = screen.getByRole("img"); // Buscamos por rol accesible
    expect(img).toHaveAttribute("src", `/heroes/${realHero.id}.jpg`); // Ruta de imagen
    expect(img).toHaveAttribute("alt", realHero.superhero); // Texto alternativo

    // 5. Verificamos que el botón de regresar está presente
    expect(
      screen.getByRole("button", { name: /regresar/i }) // Buscamos por rol y texto
    ).toBeInTheDocument();
  });

  // Prueba 2: Verifica la redirección cuando el héroe no existe
  test("debe redirigir a /marvel cuando el héroe no existe", () => {
    // Configuramos useParams para simular un ID que no existe
    useParams.mockReturnValue({ id: "heroe-inexistente" });

    // Renderizamos el componente con rutas alternativas
    render(
      <MemoryRouter initialEntries={["/hero/heroe-inexistente"]}>
        <Routes>
          <Route path="/hero/:id" element={<HeroPage />} />
          <Route path="/marvel" element={<h1>Página Marvel</h1>} />
        </Routes>
      </MemoryRouter>
    );

    // Verificamos que se muestra la página de Marvel (redirección exitosa)
    expect(screen.getByText("Página Marvel")).toBeInTheDocument();
  });

  // Prueba 3: Verifica el comportamiento del botón "Regresar"
  test("debe navegar correctamente al hacer click en Regresar", () => {
    // Obtenemos los datos reales del héroe
    const realHero = getHeroyById("dc-batman");

    // Renderizamos el componente
    render(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Routes>
          <Route path="/hero/:id" element={<HeroPage />} />
        </Routes>
      </MemoryRouter>
    );

    // Simulamos click en el botón de regresar
    const backButton = screen.getByRole("button", { name: /regresar/i });
    fireEvent.click(backButton);

    // Determinamos la ruta esperada según el publisher del héroe
    const expectedRoute = realHero.publisher.includes("Marvel")
      ? "/marvel"
      : "/dc";

    // Verificamos que navigate fue llamado con los parámetros correctos
    expect(mockUseNavigate).toHaveBeenCalledWith(expectedRoute, {
      replace: true, // Debe usar replace: true para la navegación
    });
  });
});
