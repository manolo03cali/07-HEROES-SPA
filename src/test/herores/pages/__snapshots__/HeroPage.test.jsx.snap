// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroPage /> Debe de mostrarse correctamente Snapshot 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="col-4"
    >
      <img
        alt="Batman"
        class="img-thumbnail animate__animated animate__fadeInLeft"
        src="/heroes/dc-batman.jpg"
      />
    </div>
    <div
      class="col-8"
    >
      <h3>
        Batman
      </h3>
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item"
        >
          <b>
            Alter ego:
          </b>
           
          Bruce Wayne
        </li>
        <li
          class="list-group-item"
        >
          <b>
            Publisher:
          </b>
           
          DC Comics
        </li>
        <li
          class="list-group-item"
        >
          <b>
            First appearance:
          </b>
           
          Detective Comics #27
        </li>
      </ul>
      <h5
        class="mt-3"
      >
        Characters
      </h5>
      <p>
        Bruce Wayne
      </p>
      <button
        class="btn btn-outline-primary"
      >
        Regresar
      </button>
    </div>
  </div>
</div>
`;
